### PR TITLE
[nnc] Fixed checking for loop carried dependence while fusing 2D reduction loops

### LINKED ISSUE #60310: [nnc] LoopNest::fuseLoops failed to fuse loops with reduce ops inside
## 🐛 Bug

`fuseLoops` failed to fuse loops if a reduce op exists. The problem is that for reduce ops, it detected a fake ReadAfterWrite dependence between loop iterations and therefore returned without applying `fuse` transformation. But the detection algorithm is wrong and there is no data dependence between loop iterations.

## To Reproduce

Test to reproduce the behavior:

```
  1 import torch
  2 
  3 class kernel_arena_scope(object):
  4     def __enter__(self):
  5         self.scope = torch._C._te.KernelScope()
  6 
  7     def __exit__(self, typ, val, traceback):
  8         self.scope = None
  9 
 10 with kernel_arena_scope():
 11     f32 = torch._C._te.Dtype.Float
 12     i32 = torch._C._te.Dtype.Int
 13 
 14     M, K, N = 64, 64, 64
 15     Mte, Kte, Nte = [torch._C._te.ExprHandle.int(x) for x in [M, K, N]]
 16     A = torch._C._te.Placeholder('A', f32, [Mte, Kte])
 17     B = torch._C._te.Placeholder('B', f32, [Kte, Nte])
 18 
 19     def get_dim_args(dims):
 20         dim_args = []
 21         for dim in dims:
 22             dim_args.append(torch._C._te.DimArg(dim, 'i' + str(len(dim_args))))
 23         return dim_args
 24 
 25     def compute(dims):
 26         m, n, k = dims[0], dims[1], dims[2]
 27         return A.load([m, k]) * B.load([k, n])
 28 
 29     C = torch._C._te.Reduce('C',
 30                 get_dim_args([Mte, Nte]),
 31                 torch._C._te.Sum(),
 32                 compute,
 33                 get_dim_args([Kte]),
 34                 )
 35     D = torch._C._te.Compute(
 36             'D',
 37             get_dim_args([Mte, Nte]),
 38             lambda m, n: C.load([m, n])
 39             )
 40     stmt = torch._C._te.Block([C.stmt(), D.stmt()])
 41     loopnest = torch._C._te.LoopNest(stmt, [D.buf()])
 42     print(f"original stmt: {loopnest}")
 43     cm, cn, ck = loopnest.get_loops_for(C)
 44     dm, dn = loopnest.get_loops_for(D)
 45     print(cm, dm)                                                                                                                                                             
 46     loopnest.fuse_loops([cm, dm])
 47     print(f"after fuse: {loopnest}")
```

Original stmt:
```
  for (int i0 = 0; i0 < 64; i0++) {
    for (int i1 = 0; i1 < 64; i1++) {
      C[i0, i1] = float(0);
      for (int i0_1 = 0; i0_1 < 64; i0_1++) {
        C[i0, i1] = ReduceOp((C[i0, i1]) + (A[i0, i0_1]) * (B[i0_1, i1]), reduce_args={i0});
      }
    }
  }
  for (int i0_2 = 0; i0_2 < 64; i0_2++) {
    for (int i1_1 = 0; i1_1 < 64; i1_1++) {
      D[i0_2, i1_1] = C[i0_2, i1_1];
    }
  }
```
`cm`, `dm` stmts to fuse:
```
for (int i0 = 0; i0 < 64; i0++) {
  for (int i1 = 0; i1 < 64; i1++) {
    C[i0, i1] = float(0);
    for (int i0_1 = 0; i0_1 < 64; i0_1++) {
      C[i0, i1] = ReduceOp((C[i0, i1]) + (A[i0, i0_1]) * (B[i0_1, i1]), reduce_args={i0});
    }
  }
}
```

```
for (int i0 = 0; i0 < 64; i0++) {
  for (int i1 = 0; i1 < 64; i1++) {
    D[i0, i1] = C[i0, i1];
  }
}
```

### PR DESCRIPTION
**Pull request by navahgar**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#60609**

Fixes #60310

Differential Revision: [D29386144](https://our.internmc.facebook.com/intern/diff/D29386144)

### CODE REVIEW

**Comment by navahgar** [comment]
@navahgar has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29386144).

File: test/cpp/tensorexpr/test_loopnest.cpp [hunk 1]
```diff
@@ -5868,6 +5868,163 @@ TEST(LoopNest, fuseLoopsWithReductions) {
...
+}
+
+// NOLINTNEXTLINE(cppcoreguidelines-avoid-non-const-global-variables)
+TEST(LoopNest, fuseLoopsWithComplexIndices) {
+  KernelScope kernel_scope;
+
+  // Input IR:
+  //   for (int i = 0; i < 20; i++) {
+  //     for (int j = 0; j < 20; j++) {
+  //       A[i,j*20] = i + j;
```
Comment (huiguoo) [inline test/cpp/tensorexpr/test_loopnest.cpp:5939]: A[i, j*20] -> A[i, j*20+i]

File: test/cpp/tensorexpr/test_loopnest.cpp [hunk 2]
```diff
@@ -5868,6 +5868,163 @@ TEST(LoopNest, fuseLoopsWithReductions) {
...
+  //     }
+  //   }
+  //   for (int m = 0; m < 20; m++) {
+  //     for (int n = 0; n < 20; n++) {
+  //       B[m,n] = A[m,n*20];  // Both indices of A use m
```
Comment (huiguoo) [inline test/cpp/tensorexpr/test_loopnest.cpp:5944]: A[m, n * 20] -> A[m, m * 20 + n]

File: test/cpp/tensorexpr/test_loopnest.cpp [hunk 3]
```diff
@@ -5868,6 +5868,163 @@ TEST(LoopNest, fuseLoopsWithReductions) {
...
+  VarHandle n("n", kInt);
+  auto writeA = Store::make(a_buf, {i, i * 20 + j}, i + j);
+  auto forI = For::make(i, 0, 20, For::make(j, 0, 20, writeA));
+  auto storeB = Store::make(b_buf, {m, n}, Load::make(a_buf, {m, m * 20 + n}));
+  auto forM = For::make(m, 0, 20, For::make(n, 0, 20, storeB));
+  auto par = Block::make({forI, forM});
+
+  // NOLINTNEXTLINE(cppcoreguidelines-init-variables)
+  For* fused_loop;
+  ASSERT_FALSE(LoopNest::fuseLoops({forI, forM}, &fused_loop));
```
Comment (huiguoo) [inline test/cpp/tensorexpr/test_loopnest.cpp:5961]: This case is not handled by the `equivalence` check but the indices here are equivalent, right? That is, the write and read of `A[m, m*20+n]` happens in the same iteration of the outer loop?

File: test/cpp/tensorexpr/test_loopnest.cpp [hunk 4]
```diff
@@ -5868,6 +5868,163 @@ TEST(LoopNest, fuseLoopsWithReductions) {
...
+}
+
+// NOLINTNEXTLINE(cppcoreguidelines-avoid-non-const-global-variables)
+TEST(LoopNest, fuseLoopsWithMixedLoopVarsAsIndices) {
```
Comment (huiguoo) [inline test/cpp/tensorexpr/test_loopnest.cpp:5965]: Same test as above?

**Comment by navahgar** [comment]
@navahgar has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29386144).

File: test/cpp/tensorexpr/test_loopnest.cpp (hunk 4 above)
Comment (navahgar) [reply test/cpp/tensorexpr/test_loopnest.cpp:5982]: Good catch. I had some local changes which I missed pushing here. These are actually 2 different tests, one where the loop indices are complex but the outer loop var is not used in inner dim and other where the outer loop var is also used in the inner dim.

File: test/cpp/tensorexpr/test_loopnest.cpp (hunk 3 above)
Comment (navahgar) [reply test/cpp/tensorexpr/test_loopnest.cpp]: Yes, this is one of the case where the fusion will not be allowed because we cannot prove the accesses to `A` are loop-independent after fusion.

File: test/cpp/tensorexpr/test_loopnest.cpp (hunk 2 above)
Comment (navahgar) [reply test/cpp/tensorexpr/test_loopnest.cpp]: Fixed.

File: test/cpp/tensorexpr/test_loopnest.cpp (hunk 1 above)
Comment (navahgar) [reply test/cpp/tensorexpr/test_loopnest.cpp]: Fixed

File: test/cpp/tensorexpr/test_loopnest.cpp [hunk 5]
```diff
@@ -5868,6 +5868,180 @@ TEST(LoopNest, fuseLoopsWithReductions) {
...
+  auto sumA = Store::make(
+      a_buf,
+      {i, j},
+      Add::make(Load::make(a_buf, {i, j}), Load::make(b_buf, {i, j, k})));
+  auto forK = For::make(k, 0, 100, sumA);
+  auto forJ = For::make(j, 0, 50, Block::make({initA, forK}));
+  auto forI = For::make(i, 0, 20, forJ);
+  auto storeC = Store::make(c_buf, {m, n}, Load::make(a_buf, {m, n}));
+  auto forM = For::make(m, 0, 20, For::make(n, 0, 40, storeC));
+  auto par = Block::make({forI, forM});
```
Comment (bertmaher) [inline test/cpp/tensorexpr/test_loopnest.cpp:5907]: Is there a reason to avoid using `Compute` in these tests?

**Review by bertmaher — APPROVED** [review]
I was probably being dense, but it took me a long time to understand why it's acceptable to completely ignore indices that don't involve the outer-loop variables.  If I understand correctly, it's because fusion (which changes the outer loop variables) won't change the ordering of accesses with respect to those dimensions.

I think I might be thrown by the wording of "equivalence".  Like if you're fusing loop `i`, and you're considering indices `j * 2` vs `k * 3`, those expressions don't look "equivalent" to me, but we just don't care about them.

If there's some better naming here that could help clarify it'd be great (but no need to block on this (unless the above indicates that in fact I *don't* understand the issue, in which case more explanation is probably needed :-) )).

**Comment by navahgar** [comment]
@navahgar has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29386144).

File: test/cpp/tensorexpr/test_loopnest.cpp (hunk 5 above)
Comment (navahgar) [reply test/cpp/tensorexpr/test_loopnest.cpp:5907]: Just personal preference I guess. In the past, I have had some cases where `Compute` wasn't enough to write some tests. So, I resorted to generating the IR from scratch. It helps me understand the IR better. Also, in some cases I need to refer to some intermediate loop or statements later in the test (for some checks). I can have a handle to those when I generate the tests this way.

**Comment by navahgar** [comment]
>  If I understand correctly, it's because fusion (which changes the outer loop variables) won't change the ordering of accesses with respect to those dimensions.

You are right about this.

> I think I might be thrown by the wording of "equivalence".

Yeah,  I realize that now. I have renamed the function and also updated the comments to explain things better now. Hopefully, this is more clear :). 

Thanks for pointing that out.

**Comment by navahgar** [comment]
@navahgar has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29386144).
